rumqttd/architecture.md

```diff
@@ -139,5 +139,94 @@ NOTE: Transition to ConnectionReady will schedule the connection for consumption
+
+   - `ibufs`: All incoming data to `rumqttd` (represented by a `Packet`) is stored immediately in `ibufs` (which holds a `VecDequeue`)
+   - `obufs`: All outgoing data `rumqttd` (represented by `Notification`) is stored immediately in the `obufs` (which holds a `VecDequeue`).
+
+   `obufs` also has an _in-flight queue_ to track packets that have been sent.
+
+1. Broker
+
+   Top-level entity that handles _everything_, including, but not limited to
+
+   - Config Validation
+   - Configuring and creating the Router
+   - Starting the Router
+   - Maintaining a channel that communicates with the router
+
+   When you start the Broker, the following happens ->
+
+   1. The Router is started
+   2. (on a new thread) Creates the metrics server
+   3. (on a new thread) Starts all `mqttv4` servers
+   4. (on a new thread) Starts all `mqttv5` servers
+   5. (on a new thread) Starts the prometheus listener
+   6. (on the same thread) Starts the ConsoleLink to allow
+
+2. Router
+
+   Entity managed by the broker that controls the flow of data in `rumqttd`. The router is responsible for managing, authorizing, and scheduling the flow of data between components within and connected to `rumqttd`.
+
+   <!-- At any given time, the router's event loop must choose between handling incoming messages or sending outgoing messages. -->
+
+   The router is built on the actor pattern, where it works primarily by _reacting_ to events that happen. As and when an event is added to `ibufs`, the router reacts by checking its current state (for a certain connection or message) and dispatching the appropriate action. This can involve changing other in-memory data structures, or writing the message to `obufs` to make sure it's handled by the appropriate link.
+
+   An action of any sort on the router's end which involves communication usually involves the router adding a structure or packet to one of these buffers.
+
+   The router also shares Channels, which do not contain the events themselves, but rather contain notifications to events that might be added into the shared event buffers. This also has the added benefit of preventing lock contention between different futures/tasks, which would've happened if they had to keep checking the buffers
+
+   When you start the router, the following happens ->
+
+   1. Transmission Channels get created to and from the router
+   2. A `Graveyard` to store information to persist connections
+   3. A tracker for All the alerts and metrics that are necessary to be handled
+   4. A Map between subscriptions and devices gets created
+   5. The two communication buffers `ibufs` and `obufs` get initialized
+   6. Logs for message acknowledgements, subscriptions, and alerts get created
+   7. A Packet Cache is created
+   8. The core Scheduler Object itself is created
```

Comment:
This is done when we _create_ a router, not when we start it!

> Scheduler Object

Scheduler **struct** :crab:

Reply:
I'll change that.